### PR TITLE
Add support for SSH certs

### PR DESCRIPTION
**Pull request by wiktor-k**
This doesn't work but illustrates the way parsing `KeyData` from binary AddIdentity request with cert fails.

### CODE REVIEW

**Comment by wiktor-k** [comment]
Yep, exactly. I tried to use KeyData which seem to have support for unknown types: https://github.com/wiktor-k/ssh-agent-lib/issues/3#issuecomment-2036882996

Your suggestion to use an enum may be exactly what's needed 👌 

As I mentioned in the other comment SSH binary format is not regular and one can't parse it without knowing the schema and the specs are sometimes broken (just yesterday I found reserved values I'm restrict destination constraints that were in the source but not in the spec).

**Comment by baloo** [comment]
This last commit should fix parsing, but this requires a fix upstream to expose a function:
https://github.com/RustCrypto/SSH/compare/cb2706f118c8a26b185c951ba5818abfb1352cb6...baloo:SSH:baloo/ssh-key/fixes?expand=1

File: src/proto/message.rs [hunk 1]
```diff
@@ -94,32 +98,105 @@ impl Encode for SignRequest {
...
+
+            Credential::Cert {
+                algorithm,
+                certificate,
+                privkey,
+                comment,
+            }
+        } else {
+            let algorithm = Algorithm::from_str(&alg).map_err(EncodingError::from)?;
+            let privkey = KeypairData::decode_as(reader, algorithm)?;
```
Comment (baloo) [inline src/proto/message.rs:184]: depends https://github.com/RustCrypto/SSH/pull/211

File: src/proto/privatekey.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        }
+    }
+}
+
+impl fmt::Debug for PrivateKeyData {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Dsa(_) => write!(f, "PrivateKeyData::Dsa"),
+            Self::Ecdsa(_) => write!(f, "PrivateKeyData::Ecdsa"),
+            Self::Ed25519(kp) => write!(f, "PrivateKeyData::Ed25519({kp:?})"),
```
Comment (wiktor-k) [inline src/proto/privatekey.rs:121]: I think you shouldn't print the inner type here.

File: tests/sign-and-verify.sh [hunk 3]
```diff
@@ -2,8 +2,7 @@
 
 set -euxo pipefail
 
-rm -rf ssh-agent.sock Cargo.toml.sig id_rsa id_rsa.pub agent.pub
-
+rm -rf ssh-agent.socks Cargo.toml.sig id_rsa id_rsa.pub agent.pub ca_user_key ca_user_key.pub id_rsa-cert.pub
```
Comment (wiktor-k) [inline tests/sign-and-verify.sh:5]: This looks like a typo but one sock is enough. 👍

File: tests/sign-and-verify.sh [hunk 4]
```diff
@@ -33,5 +32,15 @@ echo | ssh-add -X
...
+
+# Create and sign SSH user certificate
+# see: https://cottonlinux.com/ssh-certificates/
+echo | ssh-keygen -f ca_user_key
+ssh-keygen -t rsa -f id_rsa -N ""
+echo | ssh-keygen -s ca_user_key -I darren -n darren -V +1h -z 1 id_rsa.pub
+# Add the key with the cert
+ssh-add -t 2 id_rsa
+
 # clean up the only leftover
```
Comment (wiktor-k) [inline tests/sign-and-verify.sh:45]: This comment needs to be updated or removed.

**Review by wiktor-k — COMMENTED** [review]
This looks just great 👌 

I'll want to take a look at it on my computer but I think it's good to be merged. Great job 👏

**Comment by wiktor-k** [comment]
Okay, I've fixed a couple of small issues I've found and added test case for `AddIdentityConstrained` with multiple extensions (using my FIDO key and destination constraints). It seems to be working well.

I'll approve this PR and if @baloo doesn't have anything against I propose we merge it :)

**Comment by baloo** [comment]
ship it

**Pull request merged**